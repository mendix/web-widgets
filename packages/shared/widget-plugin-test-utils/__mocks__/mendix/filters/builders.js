exports.literal = x => ({
    type: "literal",
    value: x,
    valueType: (value => {
        if (value === undefined) {
            return "undefined";
        }
        switch (value.constructor.name) {
            case "String":
                return "String";
            case "Number":
            case "Big":
                return "Numeric";
            case "Boolean":
                return "boolean";
            case "Date":
                return "DateTime";
            case "Object":
                return "Reference";
            case "Array":
                return "ReferenceSet";
            default:
                return "undefined";
        }
    })(x)
});

exports.attribute = x => ({
    type: "attribute",
    attributeId: x
});

exports.equals = (x, y) => ({
    type: "function",
    name: "=",
    arg1: x,
    arg2: y
});

exports.notEqual = (x, y) => ({
    type: "function",
    name: "!=",
    arg1: x,
    arg2: y
});

exports.dayEquals = (x, y) => ({
    type: "function",
    name: "day:=",
    arg1: x,
    arg2: y
});

exports.dayNotEqual = (x, y) => ({
    type: "function",
    name: "day:!=",
    arg1: x,
    arg2: y
});

exports.dayGreaterThan = (x, y) => ({
    type: "function",
    name: "day:>",
    arg1: x,
    arg2: y
});

exports.dayGreaterThanOrEqual = (x, y) => ({
    type: "function",
    name: "day:>=",
    arg1: x,
    arg2: y
});

exports.dayLessThan = (x, y) => ({
    type: "function",
    name: "day:<",
    arg1: x,
    arg2: y
});

exports.dayLessThanOrEqual = (x, y) => ({
    type: "function",
    name: "day:<=",
    arg1: x,
    arg2: y
});

exports.or = (...args) => ({
    type: "function",
    name: "or",
    args
});

exports.and = (...args) => ({
    type: "function",
    name: "and",
    args
});
