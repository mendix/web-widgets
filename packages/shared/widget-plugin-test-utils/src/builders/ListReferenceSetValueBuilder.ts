import { DynamicValue, ListReferenceSetValue, ObjectItem } from "mendix";
import { dynamic } from "../primitives/dynamic.js";
import { nanoid } from "../primitives/nanoid.js";
import { objArray } from "../primitives/objArray.js";
import { Writable } from "./type-utils.js";

export class ListReferenceSetValueBuilder {
    mock: Writable<ListReferenceSetValue>;
    constructor() {
        this.mock = {
            type: "ReferenceSet",
            id: `listRefSet_${nanoid()}` as any,
            filterable: true,
            get: jest.fn((_: ObjectItem) => dynamic(objArray(1)))
        };
    }

    withGet(getter: (item: ObjectItem) => DynamicValue<ObjectItem[]>): this {
        this.mock.get = getter;
        return this;
    }

    withId(id: string): this {
        this.mock.id = id as any;
        return this;
    }

    build(): ListReferenceSetValue {
        return { ...this.mock };
    }
}
