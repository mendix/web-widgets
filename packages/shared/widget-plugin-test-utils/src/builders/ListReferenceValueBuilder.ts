import { DynamicValue, ListReferenceValue, ObjectItem } from "mendix";
import { dynamic } from "../primitives/dynamic.js";
import { nanoid } from "../primitives/nanoid.js";
import { obj } from "../primitives/obj.js";
import { Writable } from "./type-utils.js";
export class ListReferenceValueBuilder {
    mock: Writable<ListReferenceValue>;
    constructor() {
        this.mock = {
            type: "Reference",
            id: `listRef_${nanoid()}` as any,
            filterable: true,
            get: jest.fn((_: ObjectItem) => dynamic(obj()))
        };
    }

    withGet(getter: (item: ObjectItem) => DynamicValue<ObjectItem>): this {
        this.mock.get = getter;
        return this;
    }

    withId(id: string): this {
        this.mock.id = id as any;
        return this;
    }

    build(): ListReferenceValue {
        return { ...this.mock };
    }
}
