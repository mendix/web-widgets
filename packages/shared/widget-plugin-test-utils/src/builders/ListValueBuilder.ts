import type { ListValue, ObjectItem } from "mendix";
import { Status } from "../constants.js";
import { Writable } from "./type-utils.js";
import { obj } from "../primitives/obj.js";
import { objArray } from "../primitives/objArray.js";

export class ListValueBuilder {
    mock: Writable<ListValue>;
    constructor() {
        this.mock = {
            status: Status.Available,
            offset: 0,
            limit: 2,
            items: [obj(), obj()],
            totalCount: 2,
            hasMoreItems: false,
            setLimit: jest.fn(),
            setOffset: jest.fn(),
            requestTotalCount: jest.fn(),
            sortOrder: [],
            filter: undefined,
            setSortOrder: jest.fn(),
            setFilter: jest.fn(),
            reload: jest.fn()
        };
    }

    withItems(items: ObjectItem[]): this {
        this.mock.items = items;
        this.mock.totalCount = items.length;
        this.mock.limit = items.length;
        return this;
    }

    withSize(size: number): this {
        this.withItems(objArray(size));
        return this;
    }

    isLoading(): this {
        this.mock.status = Status.Loading;
        this.mock.items = undefined;
        return this;
    }

    isUnavailable(): this {
        this.mock.status = Status.Loading;
        this.mock.items = undefined;
        return this;
    }

    build(): ListValue {
        return { ...this.mock };
    }
}
