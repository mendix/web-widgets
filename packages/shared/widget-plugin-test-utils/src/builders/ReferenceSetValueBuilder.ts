import type { ObjectItem, ReferenceSetValue, Option } from "mendix";
import { Status } from "../constants.js";
import { Writable } from "./type-utils.js";

export class ReferenceSetValueBuilder {
    private readonly referenceSetValue: Writable<ReferenceSetValue> = {
        value: undefined,
        type: "ReferenceSet",
        status: Status.Available,
        validation: undefined,
        readOnly: false,
        setValidator: jest.fn(),
        setValue: jest.fn(value => this.withValue(value))
    };

    withValue(value: Option<ObjectItem[]>): ReferenceSetValueBuilder {
        this.referenceSetValue.value = value;
        return this;
    }

    isReadOnly(): ReferenceSetValueBuilder {
        this.referenceSetValue.readOnly = true;
        return this;
    }

    isLoading(): ReferenceSetValueBuilder {
        this.referenceSetValue.status = Status.Loading;
        return this.isReadOnly();
    }

    isUnavailable(): ReferenceSetValueBuilder {
        this.referenceSetValue.status = Status.Unavailable;
        return this.isReadOnly();
    }

    withValidation(validation?: string): ReferenceSetValueBuilder {
        this.referenceSetValue.validation = validation;
        return this;
    }

    build(): ReferenceSetValue {
        return this.referenceSetValue;
    }
}
