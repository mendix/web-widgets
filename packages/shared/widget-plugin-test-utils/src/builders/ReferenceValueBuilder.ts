import type { ObjectItem, ReferenceValue, Option } from "mendix";
import { Status } from "../constants.js";
import { Writable } from "./type-utils.js";

export class ReferenceValueBuilder {
    private readonly referenceValue: Writable<ReferenceValue> = {
        value: undefined,
        type: "Reference",
        status: Status.Available,
        validation: undefined,
        readOnly: false,
        setValidator: jest.fn(),
        setValue: jest.fn(value => this.withValue(value))
    };

    withValue(value: Option<ObjectItem>): ReferenceValueBuilder {
        this.referenceValue.value = value;
        return this;
    }

    isReadOnly(): ReferenceValueBuilder {
        this.referenceValue.readOnly = true;
        return this;
    }

    isLoading(): ReferenceValueBuilder {
        this.referenceValue.status = Status.Loading;
        return this.isReadOnly();
    }

    isUnavailable(): ReferenceValueBuilder {
        this.referenceValue.status = Status.Unavailable;
        return this.isReadOnly();
    }

    withValidation(validation?: string): ReferenceValueBuilder {
        this.referenceValue.validation = validation;
        return this;
    }

    build(): ReferenceValue {
        return this.referenceValue;
    }
}
