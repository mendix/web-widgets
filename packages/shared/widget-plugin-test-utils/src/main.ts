export * from "./builders/EditableValueBuilder.js";
export * from "./builders/ReferenceValueBuilder.js";
export * from "./builders/ReferenceSetValueBuilder.js";
export * from "./builders/ListAttributeValueBuilder.js";
export * from "./builders/ListValueBuilder.js";
export * from "./builders/SelectionSingleValueBuilder.js";
export * from "./builders/SelectionMultiValueBuilder.js";
export * from "./functions/cases.js";
export * from "./functions/list.js";
export * from "./functions/listAction.js";
export * from "./functions/listAttribute.js";
export * from "./functions/listExpression.js";
export * from "./functions/listRef.js";
export * from "./functions/listReferenceSet.js";
export * from "./functions/listWidget.js";
export * from "./functions/reference.js";
export * from "./functions/referenceSet.js";
export * from "./primitives/action.js";
export * from "./primitives/attrId.js";
export * from "./primitives/dynamic.js";
export * from "./primitives/editable.js";
export * from "./primitives/obj.js";
export * from "./primitives/objArray.js";
