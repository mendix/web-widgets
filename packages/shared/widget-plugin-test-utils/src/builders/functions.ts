import { ReactNode } from "react";
import type { Big } from "big.js";
import type {
    ListValue,
    ListExpressionValue,
    ListWidgetValue,
    ObjectItem,
    EditableValue,
    ListAttributeValue,
    ListActionValue,
    ActionValue,
    ReferenceValue,
    ReferenceSetValue,
    DynamicValue
} from "mendix";
import { ListValueBuilder } from "./ListValueBuilder.js";
import { ListAttributeValueBuilder } from "./ListAttributeValueBuilder.js";
import { EditableValueBuilder } from "./EditableValueBuilder.js";
import { ReferenceValueBuilder } from "./ReferenceValueBuilder.js";
import { ReferenceSetValueBuilder } from "./ReferenceSetValueBuilder.js";
import { Status } from "../constants.js";

/** @deprecated Please, uee {@link listExp} instead */
export function buildListExpression<T extends string | boolean | Date | Big>(value: T): ListExpressionValue<T> {
    return { get: () => dynamicValue<T>(value) } as Pick<ListExpressionValue, "get"> as ListExpressionValue<T>;
}

/** @deprecated Please, use {@link listWidget} instead */
export function buildWidgetValue(value: ReactNode): ListWidgetValue {
    return { get: () => value } as Pick<ListWidgetValue, "get"> as ListWidgetValue;
}

export function obj(id = Math.random().toFixed(16).slice(2, 6)): ObjectItem {
    id = `obj_${id}`;
    return { id } as unknown as ObjectItem;
}

export function objectItems(length = 1): ObjectItem[] {
    length = Math.floor(length);
    return length > 0 ? Array.from({ length }, obj) : [];
}

export function list(n: number): ListValue {
    return ListValueBuilder().withItems(objectItems(n));
}

export function listAttr<T extends string | boolean | Date | Big>(get: (item: ObjectItem) => T): ListAttributeValue<T> {
    const attr = new ListAttributeValueBuilder<T>().build();
    const attrGet = jest.fn((item: ObjectItem) =>
        new EditableValueBuilder().withValue(get(item)).build()
    ) as unknown as (item: ObjectItem) => EditableValue<T>;
    attr.get = attrGet;
    return attr;
}

export function attrId(id: string = "{unset}"): ListAttributeValue["id"] {
    return id as ListAttributeValue["id"];
}

export function listExp<T extends string | boolean | Date | Big>(get: (item: ObjectItem) => T): ListExpressionValue<T> {
    return { get: (item: ObjectItem) => dynamicValue<T>(get(item)) } as unknown as ListExpressionValue<T>;
}

export function listWidget(get: (item: ObjectItem) => ReactNode): ListWidgetValue {
    return { get } as unknown as ListWidgetValue;
}

export function listAction(customValue?: (actionFactory: typeof actionValue) => ActionValue): ListActionValue {
    return {
        get: () => (typeof customValue === "function" ? customValue(actionValue) : actionValue())
    } as unknown as ListActionValue;
}

/**
 * Short function to mock ReferenceValue.
 * @param factory - optional factory function which takes
 * ReferenceValueBuilder as first argument and returns new ReferenceValue.
 */
export function ref(factory?: (builder: ReferenceValueBuilder) => ReferenceValue): ReferenceValue {
    factory ??= builder => builder.build();
    return factory(new ReferenceValueBuilder());
}

/**
 * Short function to mock ReferenceSetValue.
 * @param factory - optional factory function which takes
 * ReferenceSetValueBuilder as first argument and returns new ReferenceValue.
 */
export function refSet(factory?: (builder: ReferenceSetValueBuilder) => ReferenceSetValue): ReferenceSetValue {
    factory ??= builder => builder.build();
    return factory(new ReferenceSetValueBuilder());
}

export function dynamicValue<T>(value?: T, loading?: boolean): DynamicValue<T> {
    if (loading) {
        return { status: Status.Loading, value };
    }
    return value !== undefined ? { status: Status.Available, value } : { status: Status.Unavailable, value: undefined };
}

export function actionValue(canExecute = true, isExecuting = false): ActionValue {
    return { canExecute, isExecuting, execute: jest.fn() };
}
