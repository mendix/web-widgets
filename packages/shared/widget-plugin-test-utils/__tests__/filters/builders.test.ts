jest.mock("mendix/filters/builders");
import { Big } from "big.js";
import {
    literal,
    attribute,
    equals,
    or,
    and,
    dayLessThanOrEqual,
    dayLessThan,
    dayGreaterThanOrEqual
} from "mendix/filters/builders";
import * as builders from "mendix/filters/builders";
import { attrId, obj } from "../../src/main";

describe("mendix/filters/builders mock", () => {
    test.each([
        [true, "boolean"],
        [false, "boolean"],
        ["!", "String"],
        ["", "String"],
        [new Big(0), "Numeric"],
        [new Big(0.49), "Numeric"],
        [new Big(-1), "Numeric"],
        [new Date("2024-09-20"), "DateTime"],
        [obj("12345"), "Reference"],
        [[obj("12345")], "ReferenceSet"],
        [undefined, "undefined"]
    ])("literal(%j) returns object with value type '%s'", (value, expectedType) => {
        const output = literal(value);
        expect(output.type).toBe("literal");
        expect(output.value).toBe(value);
        expect(output.valueType).toBe(expectedType);
    });

    test("attribute", () => {
        expect(attribute(attrId("csj_5"))).toEqual({
            type: "attribute",
            attributeId: "attr_csj_5"
        });
    });

    test.each([
        ["equals", "="],
        ["notEqual", "!="],
        ["dayEquals", "day:="],
        ["dayNotEqual", "day:!="],
        ["dayGreaterThan", "day:>"],
        ["dayGreaterThanOrEqual", "day:>="],
        ["dayLessThan", "day:<"],
        ["dayLessThanOrEqual", "day:<="]
    ])("%s returns correct shape", (fn, name) => {
        expect(
            (builders[fn as keyof typeof builders] as any)(attribute(attrId("foo_bar")), literal(undefined))
        ).toEqual({
            type: "function",
            name,
            arg1: {
                type: "attribute",
                attributeId: "attr_foo_bar"
            },
            arg2: {
                type: "literal",
                value: undefined,
                valueType: "undefined"
            }
        });
    });

    test("or", () => {
        expect(
            or(
                equals(literal(true), literal(true)),
                equals(literal("!"), literal("!")),
                equals(literal("X"), literal(false))
            )
        ).toEqual({
            type: "function",
            name: "or",
            args: [
                equals(literal(true), literal(true)),
                equals(literal("!"), literal("!")),
                equals(literal("X"), literal(false))
            ]
        });
    });

    test("and", () => {
        expect(
            and(
                equals(literal(true), literal(true)),
                equals(literal("!"), literal("!")),
                equals(literal("X"), literal(false))
            )
        ).toEqual({
            type: "function",
            name: "and",
            args: [
                equals(literal(true), literal(true)),
                equals(literal("!"), literal("!")),
                equals(literal("X"), literal(false))
            ]
        });
    });

    describe("nested condition", () => {
        test("case 1", () => {
            expect(
                or(
                    dayLessThanOrEqual(attribute(attrId("jei_5")), literal(new Date("1961-04-12T23:00:00.000Z"))),
                    dayLessThanOrEqual(attribute(attrId("jei_6")), literal(new Date("1961-04-12T23:00:00.000Z")))
                )
            ).toMatchSnapshot();
        });

        test("case 2", () => {
            expect(
                and(
                    dayGreaterThanOrEqual(attribute(attrId("jei_5")), literal(new Date("1961-04-12T23:00:00.000Z"))),
                    dayLessThan(attribute(attrId("jei_5")), literal(new Date("1961-04-12T23:00:00.000Z")))
                )
            ).toMatchSnapshot();
        });
    });
});
