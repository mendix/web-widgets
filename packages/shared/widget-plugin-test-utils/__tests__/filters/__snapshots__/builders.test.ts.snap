// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mendix/filters/builders mock nested condition case 1 1`] = `
{
  "args": [
    {
      "arg1": {
        "attributeId": "attr_jei_5",
        "type": "attribute",
      },
      "arg2": {
        "type": "literal",
        "value": 1961-04-12T23:00:00.000Z,
        "valueType": "DateTime",
      },
      "name": "day:<=",
      "type": "function",
    },
    {
      "arg1": {
        "attributeId": "attr_jei_6",
        "type": "attribute",
      },
      "arg2": {
        "type": "literal",
        "value": 1961-04-12T23:00:00.000Z,
        "valueType": "DateTime",
      },
      "name": "day:<=",
      "type": "function",
    },
  ],
  "name": "or",
  "type": "function",
}
`;

exports[`mendix/filters/builders mock nested condition case 2 1`] = `
{
  "args": [
    {
      "arg1": {
        "attributeId": "attr_jei_5",
        "type": "attribute",
      },
      "arg2": {
        "type": "literal",
        "value": 1961-04-12T23:00:00.000Z,
        "valueType": "DateTime",
      },
      "name": "day:>=",
      "type": "function",
    },
    {
      "arg1": {
        "attributeId": "attr_jei_5",
        "type": "attribute",
      },
      "arg2": {
        "type": "literal",
        "value": 1961-04-12T23:00:00.000Z,
        "valueType": "DateTime",
      },
      "name": "day:<",
      "type": "function",
    },
  ],
  "name": "and",
  "type": "function",
}
`;
